### PR TITLE
Configure validation frequency

### PR DESCRIPTION
**Pull request by JohnGiorgi**
Fixes #5214 .

Changes proposed in this pull request:

- Add a boolean flag to `GradientDescentTrainer` (`_should_validate_this_epoch`) that controls whether validation will occur at the end of an epoch.
- Add a callback, `ShouldValidateCallback` that allows a user to configure validation frequency.
- The callback address both points in #5214, by allowing you to delay validation until a certain number of epochs have elapsed (with `validation_start`) and the validation frequency (with `validation_interval`).

## Before submitting

- [X] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [x] All GitHub Actions jobs for my pull request have passed.
- [x] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
@epwalsh Mind taking a look when you get a chance? This is a first attempt to address both points made in #5214!

File: allennlp/training/callbacks/should_validate.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+if TYPE_CHECKING:
+    from allennlp.training.gradient_descent_trainer import GradientDescentTrainer
+
+
+@TrainerCallback.register("should_validate_callback")
+class ShouldValidateCallback(TrainerCallback):
+    """
+    A callback that you can pass to the `GradientDescentTrainer` to change the frequency of
+    validation during training. If `start_validation` is not `None`, validation will not occur until
```
Comment (epwalsh) [inline allennlp/training/callbacks/should_validate.py:13]: I think `validation_start` sounds a little better than `start_validation`

File: tests/training/trainer_test.py [hunk 2]
```diff
@@ -1331,6 +1332,44 @@ def test_console_log_callback(self):
...
+        callback.on_epoch(trainer, metrics={}, epoch=1)
+        assert not trainer._should_validate_this_epoch
+
+        # Satisfies 'start_validation' but not 'validation_interval'
+        callback.on_epoch(trainer, metrics={}, epoch=2)
+        assert not trainer._should_validate_this_epoch
+
+        # Satisfies both 'start_validation' and 'validation_interval'
+        callback.on_epoch(trainer, metrics={}, epoch=4)
+        assert trainer._should_validate_this_epoch
```
Comment (epwalsh) [inline tests/training/trainer_test.py:1371]: I think it would be better to test that the training loop actually runs when using this callback.

**Review by epwalsh — CHANGES_REQUESTED** [review]
Thanks @JohnGiorgi, this seems like a good approach. I have a couple comments 👇

File: allennlp/training/callbacks/should_validate.py (hunk 1 above)
Comment (JohnGiorgi) [reply allennlp/training/callbacks/should_validate.py]: Changed! Yeah, I was struggling with the naming, if you have any other name change suggestions let me know!

File: tests/training/trainer_test.py (hunk 2 above)
Comment (JohnGiorgi) [reply tests/training/trainer_test.py]: I tweaked the test so that `trainer.train()` is called with the callback provided. I could split this into its own test though.

File: allennlp/training/gradient_descent_trainer.py [hunk 3]
```diff
@@ -811,7 +812,7 @@ def _try_train(self) -> Tuple[Dict[str, Any], int]:
                     metrics["peak_" + key] = max(metrics.get("peak_" + key, 0), value)
 
             this_epoch_val_metric: float = 0.0
```
Comment (epwalsh) [inline allennlp/training/gradient_descent_trainer.py:814]: I'm a little concerned about leaving `this_epoch_val_metric` set at 0.0 when validation is skipped. But I think we could make it `None` by default. `LearningRateScheduler.step()` and `MomentumScheduler.step()` (both called below with `this_epoch_val_metric`) take an `Optional[float]`, so that should be okay there.

```suggestion
            this_epoch_val_metric: Optional[float] = None
```

We should also be careful where `metric_track.is_best_so_far()` is checked. Maybe we only check `.is_best_so_far` when `should_validate_this_epoch` is `True`.

Comment (JohnGiorgi) [reply allennlp/training/gradient_descent_trainer.py]: What if `Model.is_best_so_far()` returned `False` if `_should_validate_this_epoch` is `False`?

Comment (epwalsh) [reply allennlp/training/gradient_descent_trainer.py]: That would work too I guess, but how would `MetricTracker` know? Would the trainer set `metric_tracker._is_best_so_far` to `False` when `_should_validate_this_epoch` is `False`? It just seems cleaner for the trainer to avoid checking the metric tracker when `_should_validate_this_epoch` is `False`.

Comment (JohnGiorgi) [reply allennlp/training/gradient_descent_trainer.py]: Okay, I changed the trainer so that `should_validate_this_epoch` is checked everywhere `is_best_so_far()` is called (I count three if statements). Is that what you had in mind?

File: allennlp/training/gradient_descent_trainer.py [hunk 4]
```diff
@@ -949,7 +954,9 @@ def _try_train(self) -> Tuple[Dict[str, Any], int]:
             epoch = self._num_epochs - 1
 
         # Load the best model state before returning
-        if self._best_model_filename is None or self._metric_tracker.is_best_so_far():
+        if self._should_validate_this_epoch and (
+            self._best_model_filename is None or self._metric_tracker.is_best_so_far()
+        ):
```
Comment (JohnGiorgi) [inline allennlp/training/gradient_descent_trainer.py:959]: I most unsure about this change

File: allennlp/training/gradient_descent_trainer.py (hunk 3 above)
Comment (epwalsh) [reply allennlp/training/gradient_descent_trainer.py]: Right, yeup!

File: allennlp/training/gradient_descent_trainer.py (hunk 4 above)
Comment (epwalsh) [reply allennlp/training/gradient_descent_trainer.py:959]: Hmm I think we can omit the check here, i.e. leave it as it was.

**Review by epwalsh — APPROVED** [review]
LGTM! Thanks @JohnGiorgi!

**Comment by JohnGiorgi** [comment]
> LGTM! Thanks @JohnGiorgi!

Cool! Thanks @epwalsh
